### DISCUSSION TITLE
API Deprecation Policy

### DISCUSSION THREAD

**Discussion by YangKeao** [chaos-mesh/chaos-mesh#2805] [General]
I have started to take a look into the multi-version api, which will enable us to modify our API without fear to break the users. However, leaving aside the technical things, the strategy always needs more discussion. How should we promote new API and deprecate the older one? How long is the expected lifecycle of an API?

A good reference is from kubernetes: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

But we can not just obey the kubernetes' way. There are so many differences between us and kubernetes. For example:

1. All apis (at least all chaos) are in a single group, so they shouldn't/couldn't be moved to a "stabler" version at the same time. For example, the `NetworkChaos` is ready for production, while the `IOChaos` and `HTTPChaos` seem worse.
2. If we allow multiple versions for one group have different contents (e.g. only `PodChaos` and `NetworkChaos` exist in `v1`), how should we handle the dependencies? For example, the field dependency (`v1alpha2/NetworkChaos` depend on `v1alpha1/PodSelector`), the resource dependency (`v1alpha1/Schedule` could depend on `v1alpha2/NetworkChaos`). The loop dependency is really a problem, considering the conversion.

**Comment by STRRL** [comment]
+1 

Maybe we also need to declare with more detail with lifecycle like: "experimental" -> "stable" -> "deprecated".

**Reply by YangKeao** [reply]
If every object is in the same group, it cannot be split into multiple phases. For example, if Chaos Mesh has come to `v1`, but we want to add a new object (like `SOMEChaos`), and want it to be experimental, what should we do?

A better way to do so is to split them into multiple groups, but the migration could be an awful experience (again).
